compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPackSharedMemoryAlloc.cpp

```diff
@@ -0,0 +1,217 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Analysis/Liveness.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/NVGPU/IR/NVGPUDialect.h"
+#include "mlir/IR/Dominance.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-pack-shared-memory-alloc"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Sink allocs in the CFG to reduce its liverange.
+static void sinkAlloc(const SmallVector<Operation *> &allocs,
+                      DominanceInfo &dominators) {
+  for (Operation *sinkOp : allocs) {
+    Block *dom = nullptr;
+    for (Operation *user : sinkOp->getUsers()) {
+      if (!dom) {
+        dom = user->getBlock();
+        // Find the block in the same region.
+        while (dom->getParent() != sinkOp->getParentRegion()) {
+          dom = dom->getParentOp()->getBlock();
+        }
+        continue;
+      }
+      dom = dominators.findNearestCommonDominator(dom, user->getBlock());
+    }
+    llvm::SmallDenseSet<Operation *> users;
+    for (Operation *user : sinkOp->getUsers()) {
+      while (user->getParentRegion() != sinkOp->getParentRegion()) {
+        user = user->getParentOp();
+      }
+      users.insert(user);
+    }
+    Operation *firstUse = dom->getTerminator();
+    for (Operation &op : dom->getOperations()) {
+      if (users.count(&op)) {
+        firstUse = &op;
+        break;
+      }
+    }
+    sinkOp->moveBefore(firstUse);
+  }
+}
+
+/// Analyze the liverange of allocs and set them in individual groups if
+/// possible and mark allocs with an attributes.
+/// The algorithm is a simplistic memory allocation solution. It sorts
+/// allocations into alias groups. Everytime two alloc's liverange interfers
+/// they are merge into the same group. If a new alloc is part of multiple alias
+/// groups all those are merged into one. At the end we are left with groups of
+/// allocations that are disjoint and can use the same memory.
+// TODO: Move this to a common place if needed by other backends.
+static void analyzeSharedMemoryAlloc(
+    func::FuncOp funcOp, const SmallVector<Operation *> &allocs,
+    SmallVector<SmallVector<Operation *>> &aliasGroups) {
+  struct AllocGroup {
+    SmallVector<Operation *> allocs;
+    // Keep track of every operation where any of the alloc in the group is
+    // live.
+    llvm::DenseSet<Operation *> liveness;
+  };
+  Liveness liveness(funcOp);
+  SmallVector<AllocGroup> groups;
+  for (Operation *alloc : allocs) {
```

Comment:
`llvm::Equivalence` is slightly different than what I need here and it makes things a bit less intuitive for this simple use case. I added some extra comments and added `AliasGroup`. Hopefully it should be easier to read. Let me know what you think.

Reply:
Could you explain why it is different from a union-find data structure (it reads almost exactly as that)

Reply:
Upsignalling this comment again. Still having a hard time navigating the logic of the custom data structures. Can you explain a bit more why this is not an `llvm::Equivalence` class. Advantage of using existing data structures is it makes the logic clearer to read (I have read through this a couple of times, but without knowing what the data structures are holding, its hard to follow)

Reply:
Sorry I had missed the second comment. `llvm::Equivalence` doesn't help much here because I still need to be able to track the livness of the group of allocs as one to be able to have decent use of the data structure. Therefore I never need to use the data structure to look for a node. The concept of leaders also doesn't help much here for the same reason. I don't think use `llvm::Equivalence` to signal that this is liveranges is worth the extra code needed to switch to this data structure, I would mostly use the data structure to merge loop through the groups.
I added some comments that hopefully make the idea of the tracking more obvious.